File: Stack/stack.m
//
//  Stack.m
//  dsa
//
//  Created by Brian Tamsing on 1/25/22.
//

#import "Stack.h"

@interface Stack ()

@property (strong, nonatomic) NSMutableArray<id> *arr;

@end

@implementation Stack

- (instancetype)init {
    if (self = [super init]) {
        _arr = [[NSMutableArray alloc] init];
    }
    
    return self;
}

- (NSUInteger)count {
    return [self.arr count];
}

- (BOOL)isEmpty {
    return [self.arr count] == 0;
}

- (void)push:(id)val {
    [self.arr addObject:val];
}

- (id)pop {
    id val = [self.arr lastObject];
    [self.arr removeLastObject];
    
    return val;
}

- (id)peek {
    return [self.arr lastObject];
}

@end

File: Stack/Swift/stack.swift
//
//  Stack.swift
//  dsa
//
//  Created by Brian Tamsing on 10/16/21.
//

import Foundation

struct Stack<Element> {
    private var array = [Element]()
    
    var isEmpty: Bool {
        return array.isEmpty
    }
    
    var count: Int {
        return array.count
    }
    
    mutating func push(_ element: Element) {
        array.append(element)
    }
    
    mutating func pop() -> Element? {
        return array.popLast()
    }
    
    func peek() -> Element? {
        return array.last
    }
}